cyto_dl/callbacks/latent_walk_diffae.py

```diff
@@ -0,0 +1,127 @@
+        font = cv2.FONT_HERSHEY_SIMPLEX
+        font_scale = 0.5
+        color = tuple([img.max()]*3) 
+        thickness = 1
+        text_size = cv2.getTextSize(text, font, font_scale, thickness)[0]
+        text_x = img.shape[1] - text_size[0] - 3  # 3 pixels from the right edge
+        text_y = text_size[1] + 3  # 3 pixels from the top edge
+        cv2.putText(img, text, (text_x, text_y), font, font_scale, color, thickness)
+        return img
+
+    def _write_pc_vals(self, walk_img, ranges):
+        """
+        Write PC index and value on image
+        """
+        idx = 0
+        for i, range_ in enumerate(ranges):
+            for val in range_:
+                walk_img[idx]= self._write_text(walk_img[idx], f"PC{i+1}:{val:.1f}")
+                idx += 1
+        return walk_img
+
+    def _latent_walk(self, feats, model, save_path):
+        # catch if only one batch for validation
+        if len(feats.shape) == 1 or feats.shape[0] < self.num_pcs:
+            warn(f"Insufficient data for latent walk with {self.num_pcs} PCs. Skipping...")
+            return
+        pca_data = self.pca.fit_transform(feats)
+        print(f"Explained variance ratio: {self.pca['pca'].explained_variance_ratio_}")
+        walk = []
+        ranges = []
+        for pc in np.arange(self.num_pcs):
+            std = pca_data[:, pc].std()
+            if self.sigma_range is None:
+                min = pca_data[:, pc].min() / std
+                max = pca_data[:, pc].max() / std
+                range_ = np.linspace(min, max, self.n_steps)
+            else:
+                range_ = np.arange(-self.sigma_range, self.sigma_range + 0.01)
+            print(f"PC{pc} range: {range_}")
+            for i in range_:
+                array = np.zeros(self.num_pcs)
+                array[pc] = i * std
+                walk.append(array)
+            ranges.append(range_)
+        walk = np.stack(walk).squeeze()
+        walk = self.pca.inverse_transform(walk)
+        walk = torch.from_numpy(walk).float().to(model.device)
+        walk_img = model.generate_from_latent(walk, n_noise_samples=self.n_noise_samples, average=self.average, save=False, batch_size=self.batch_size)
+        walk_img = self._write_pc_vals(walk_img, ranges)
+        OmeTiffWriter.save(uri=save_path, data=walk_img)
+
+    def on_validation_batch_end(
+        self, trainer, pl_module, outputs, batch, batch_idx, dataloader_idx=0
+    ):
+        if (trainer.current_epoch + 1) % self.every_n_epoch == 0:
+            latent_feat = detach(outputs[1].squeeze())
+            self.val_feats.append(latent_feat)
+
+    def on_validation_epoch_end(self, trainer, pl_module):
+        if (trainer.current_epoch + 1) % self.every_n_epoch == 0:
+            # aggregate all latent features for PCA
+            feats = np.concatenate(self.val_feats)
+            self._latent_walk(feats, trainer.model, f"{pl_module.hparams.save_dir}/{trainer.current_epoch+1}_latent_walk.tiff")
```

Comment:
so this latent walk is only for validation features? is that helpful?

Reply:
And prediction. It helps with pc visualization during training and understanding when the conditioning is being used. Are there other places where you'd want to see it?